src/storage/src/render/upsert.rs

```diff
@@ -157,17 +157,29 @@ pub fn rehydration_finished<G, T>(
 
     builder.build(move |_capabilities| async move {
         loop {
-            if let Some(AsyncEvent::Progress(frontier)) = input.next().await {
-                if PartialOrder::less_equal(&resume_upper, &frontier) {
-                    tracing::info!(
-                        "timely-{} source {} reached the resume upper ({:?}) across all workers",
+            match input.next().await {
+                Some(AsyncEvent::Progress(frontier)) => {
+                    if PartialOrder::less_equal(&resume_upper, &frontier) {
+                        tracing::info!(
+                        "timely-{} upsert source {} has downgraded past the resume upper ({:?}) across all workers",
                         worker_id,
                         id,
                         resume_upper
                     );
+                        drop(token);
+                        break;
+                    }
+                }
+                None => {
+                    // Shutdown has been triggered or the shard is closed, so we shutdown.
+                    // Note that we will likely get a `Progress([])` event before this,
+                    // but we cover it for an abundance of caution.
                     drop(token);
                     break;
                 }
+                _ => {
+                    // we don't expect data
```

Comment:
nit: should we log places like here just in case?

Reply:
nah, its `Infallible` anyways